x-pack/plugins/fleet/public/applications/fleet/sections/agents/agent_list_page/components/tags_add_remove.tsx

```diff
@@ -0,0 +1,131 @@
+/*
+ * Copyright Elasticsearch B.V. and/or licensed to Elasticsearch B.V. under one
+ * or more contributor license agreements. Licensed under the Elastic License
+ * 2.0; you may not use this file except in compliance with the Elastic License
+ * 2.0.
+ */
+
+import React, { Fragment, useCallback, useEffect, useState } from 'react';
+import { i18n } from '@kbn/i18n';
+import type { EuiSelectableOption } from '@elastic/eui';
+import { EuiButtonEmpty, EuiIcon, EuiSelectable, EuiWrappingPopover } from '@elastic/eui';
+import { FormattedMessage } from '@kbn/i18n-react';
+
+import { sendPutAgentTagsUpdate, useStartServices } from '../../../../hooks';
+
+interface Props {
+  agentId: string;
+  allTags: string[];
+  selectedTags: string[];
+  button: HTMLElement;
+  onTagsUpdated: () => void;
+}
+
+export const TagsAddRemove: React.FC<Props> = ({
+  agentId,
+  allTags,
+  selectedTags,
+  button,
+  onTagsUpdated,
+}: {
+  agentId: string;
+  allTags: string[];
+  selectedTags: string[];
+  button: HTMLElement;
+  onTagsUpdated: () => void;
+}) => {
+  const labelsFromTags = useCallback(
+    (tags: string[]) =>
+      tags.map((tag: string) => ({
+        label: tag,
+        checked: selectedTags.includes(tag) ? 'on' : undefined,
+      })),
+    [selectedTags]
+  );
+
+  const { notifications } = useStartServices();
+  const [labels, setLabels] = useState<Array<EuiSelectableOption<any>>>(labelsFromTags(allTags));
+  const [searchValue, setSearchValue] = useState<string | undefined>(undefined);
+  const [isPopoverOpen, setIsPopoverOpen] = useState(true);
+  const closePopover = () => setIsPopoverOpen(false);
+
+  // update labels after tags changing
+  useEffect(() => {
+    setLabels(labelsFromTags(allTags));
+  }, [allTags, labelsFromTags]);
+
+  const updateTags = async (newTags: string[]) => {
+    const res = await sendPutAgentTagsUpdate(agentId, { tags: newTags });
+
+    if (res.error) {
+      const errorMessage = i18n.translate('xpack.fleet.updateAgentTags.errorNotificationTitle', {
+        defaultMessage: 'Tags update failed',
+      });
+      notifications.toasts.addError(res.error, { title: errorMessage });
+      throw res.error;
```

Comment:
this will cause unhandled error rejection, I think it make sense to put the whole thing into a try catch and on error show the notifications toasts